src/cockpit/superuser.py

```diff
@@ -0,0 +1,215 @@
+
+import asyncio
+import logging
+import os
+import pwd
+
+from systemd_ctypes import bus
+
+from typing import Dict, Optional
+
+from .router import Router, RoutingError, RoutingRule
+from .peer import Peer, PeerStateListener
+
+logger = logging.getLogger(__name__)
+
+SUPERUSER_AUTH_COOKIE = 'supermarius'
+
+# Keep a static list, for the time being.
+SUPERUSER_BRIDGES = {
+    'sudo': (
+        ['sudo', '--askpass', 'cockpit-bridge', '--privileged'], {
+            'SUDO_ASKPASS': '/usr/libexec/cockpit-askpass'
+        }
+    ),
+}
+
+
+class SuperuserStartup:
+    peer: Peer  # the peer that's being started
+
+    def success(self, rule: SuperuserRoutingRule) -> None:
+        raise NotImplementedError
+
+    def failed(self, rule: SuperuserRoutingRule, exc: Exception) -> None:
+        raise NotImplementedError
+
+    def auth(self, rule: SuperuserRoutingRule, prompt: str, echo: bool) -> None:
+        raise NotImplementedError
+
+
+class ControlMessageStartup(SuperuserStartup):
+    def success(self, rule: SuperuserRoutingRule) -> None:
+        rule.router.write_control(command='superuser-init-done')
+
+    def failed(self, rule: SuperuserRoutingRule, exc: Exception) -> None:
+        rule.router.write_control(command='superuser-init-done')
+
+    def auth(self, rule: SuperuserRoutingRule, prompt: str, echo: bool) -> None:
+        username = pwd.getpwuid(os.getuid()).pw_name
+        hexuser = ''.join(f'{c:02x}' for c in username.encode('ascii'))
+        rule.router.write_control(command='authorize', cookie=SUPERUSER_AUTH_COOKIE, challenge=f'plain1:{hexuser}')
+
+
+class DBusStartup(SuperuserStartup):
+    future: asyncio.Future
+
+    def __init__(self):
+        self.future = asyncio.get_running_loop().create_future()
+
+    def success(self, rule: SuperuserRoutingRule) -> None:
+        self.future.set_result(None)
+
+    def failed(self, rule: SuperuserRoutingRule, exc: Exception) -> None:
+        self.future.set_exception(exc)
+
+    def auth(self, rule: SuperuserRoutingRule, prompt: str, echo: bool) -> None:
+        rule.prompt('', prompt, '', echo, '')
+
+    async def wait(self) -> None:
+        await self.future
+
+
+class SuperuserRoutingRule(PeerStateListener, RoutingRule, bus.Object, interface='cockpit.Superuser'):
+    startup: Optional[SuperuserStartup]
+    peer: Optional[Peer]
+
+    # D-Bus signals
+    prompt = bus.Interface.Signal('s', 's', 's', 'b', 's')  # message, prompt, default, echo, error
+
+    # D-Bus properties
+    bridges = bus.Interface.Property('as', value=[])
+    current = bus.Interface.Property('s', value='none')
+
+    # RoutingRule
+    def apply_rule(self, options: Dict[str, object]) -> Optional[Peer]:
+        if not options.get('superuser') or self.current == 'root':
+            # superuser not requested, or already superuser?  Next rule.
+            return None
+        elif self.peer:
+            # superuser requested and active?  Return it.
+            return self.peer
+        else:
```

Comment:
There is also `superuser: "try"`.

Reply:
I added this.  It's remarkably simple.  In case "try" is given, `return self.peer`.  then it will either be the root bridge, or None, which is what we want.